# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/power/WakeLockJS.cpp
# Commit: 1c521c653820
# Full Hash: 1c521c653820dded2f57732401be0bd508ebdf89
# Author: Edgar Chen <echen@mozilla.com>
# Date: 2023-12-15 15:59:25
# Description:
#   Bug 1869118 - Fix crash in WakeLockJS::Observe; r=dom-core,CanadaHonk
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D196487
# ==============================================================================

diff -r 1f2343c3142b -r 1c521c653820 dom/power/WakeLockJS.cpp
--- a/dom/power/WakeLockJS.cpp	Fri Dec 15 06:58:12 2023 +0000
+++ b/dom/power/WakeLockJS.cpp	Fri Dec 15 07:08:17 2023 +0000
@@ -110,7 +110,17 @@
   MOZ_LOG(sLogger, LogLevel::Debug, ("Released wake lock sentinel"));
 }
 
-NS_IMPL_CYCLE_COLLECTION_WRAPPERCACHE(WakeLockJS, mWindow)
+NS_IMPL_CYCLE_COLLECTION_WRAPPERCACHE_CLASS(WakeLockJS)
+
+NS_IMPL_CYCLE_COLLECTION_TRAVERSE_BEGIN(WakeLockJS)
+  NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mWindow)
+NS_IMPL_CYCLE_COLLECTION_TRAVERSE_END
+
+NS_IMPL_CYCLE_COLLECTION_UNLINK_BEGIN(WakeLockJS)
+  tmp->DetachListeners();
+  NS_IMPL_CYCLE_COLLECTION_UNLINK(mWindow)
+  NS_IMPL_CYCLE_COLLECTION_UNLINK_PRESERVED_WRAPPER
+NS_IMPL_CYCLE_COLLECTION_UNLINK_END
 
 NS_IMPL_CYCLE_COLLECTING_ADDREF(WakeLockJS)
 NS_IMPL_CYCLE_COLLECTING_RELEASE(WakeLockJS)
